gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/JobInfo.java

```diff
@@ -0,0 +1,305 @@
+/*
+ * Copyright 2015 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.gcloud.bigquery;
+
+import com.google.api.services.bigquery.model.Job;
+import com.google.common.base.Function;
+import com.google.common.base.MoreObjects;
+
+import java.io.Serializable;
+import java.util.Objects;
+
+/**
+ * Base class for Google BigQuery Job information. Jobs are objects that manage asynchronous tasks
+ * such as running queries, loading data, and exporting data. Use {@link CopyJobInfo} for a job that
+ * copies an existing table. Use {@link ExtractJobInfo} for a job that exports a table to Google
+ * Cloud Storage. Use {@link LoadJobInfo} for a job that loads data from Google Cloud Storage into
+ * a table. Use {@link QueryJobInfo} for a job that runs a query.
+ */
+public abstract class JobInfo<StatisticsT extends JobStatistics> implements Serializable {
+
+  static final Function<Job, JobInfo> FROM_PB_FUNCTION =
+      new Function<Job, JobInfo>() {
+        @Override
+        public JobInfo apply(Job pb) {
+          return JobInfo.fromPb(pb);
+        }
+      };
+  private static final long serialVersionUID = -7086529810736715842L;
+
+  public enum CreateDisposition {
+    CREATE_IF_NEEDED,
+    CREATE_NEVER
+  }
+
+  public enum WriteDisposition {
+    WRITE_TRUNCATE,
+    WRITE_APPEND,
+    WRITE_EMPTY
+  }
+
+  private final String etag;
+  private final String id;
+  private final JobId jobId;
+  private final String selfLink;
+  private final JobStatus status;
+  private final StatisticsT statistics;
+  private final Boolean dryRun;
+  private final String userEmail;
+
+  public static abstract class Builder<T extends JobInfo, StatisticsT extends JobStatistics,
```

Comment:
We should be consistent with generic type naming. I would either use I, S and B or JobInfoT, StatisticsT and BuilderT.